### DISCUSSION TITLE
SLAM-VI on MuSHR, looking for advices

### DISCUSSION THREAD

**Discussion by Tomlmgh** [prl-mushr/mushr#94] [Show and tell]
Hello everyone ! We are a team of students working with a company trying to implement a real time SLAM-VI algorithm in a MuSHR robot to try to advance research in this field.
We want to test implementations on the simulator, and then on the robot (we don't often have access to the robot).

So far we have done a state of the art of some SLAM Algorithm we found and **we are looking for the help of the MuSHR community** to give us your opinions on the algorithm we chose, the feasibility of the implementation **on the simu and the robot**

Here is our state of the art :

![state1](https://user-images.githubusercontent.com/58824093/205872857-c8e9a735-2240-4a7a-94eb-9a4e42529aef.PNG)
![state2](https://user-images.githubusercontent.com/58824093/205872877-664bd427-a326-47ef-af9e-4bbdcf7b2c40.PNG)

The first 3 are those that seems to be the best to try (for us), ROVIOLI and Vins Fusion seems fine but we don't have time to try many algorithm so we want to focus on the most promising ones, and we exclude the last ones because of the reason listed on the state of the art.

If you have any tips, advice or if you tried something similar we will be happy to welcome them !

Thank you very much for your help.

@tomlmgh @SashelI @mgodineau @lilixe @Matoufk 
@Arbalestrille

**Comment by sidtalia** [comment]
Hi, we have found in general is that the MuSHR's current compute capability (Jetson nano) is not particularly suitable for Visual odometry/SLAM. The best FPS we could get for rgbd_odometry with no loop-closure or mapping was 15-20 fps with less than 150 features being tracked. We are currently working on a newer version of MuSHR and we recommend using a fair amount of compute power for real-time VI-SLAM (Jetson AGX Xavier or Orin is recommended, but an i7 processor with good amount of RAM and GPU resources would be ideal if your research allows for it). You can always use the MuSHR for collecting sensor data and running the VI-slam on a desktop if you don't need a real-time system.

That being said, while we haven't tested all of the frameworks listed by you, we have tried using [rtabmap-ros](http://wiki.ros.org/rtabmap_ros#Visual_and_Lidar_Odometry) package for Visual odometry (which internally implements something similar to ORB-SLAM except that the feature detector is interchangeable with FAST/BRIEF /SIFT/SURF/ORB and so on). 
We tried VINS-Fusion (RGB-D I think) but I believe it had some issues with calibration/initialization which have been faced by [others](https://github.com/HKUST-Aerial-Robotics/VINS-Fusion/issues?q=realsense) as well.
Here's a [basic launch file](https://gist.github.com/naughtyStark/9345351fc163c0507adc8cd8e85829ed) that uses rtabmap's rgb-d VO and here is the corresponding [realsense launch file](https://gist.github.com/naughtyStark/de431b70f4017fc8dfd8e6c993554112).
I believe rtabmap ros also allows running other state of the art visual odometry frameworks but we haven't really looked into it. Some useful documentation for that has been provided [here](http://wiki.ros.org/rtabmap_ros/Tutorials/StereoOutdoorMapping).

Let me know if you have further questions!
Sidharth Talia.

**Reply by SashelI** [reply]
Thanks.

> Hi, we have found in general is that the MuSHR's current compute capability (Jetson nano) is not particularly suitable for Visual odometry/SLAM. The best FPS we could get for rgbd_odometry with no loop-closure or mapping was 15-20 fps with less than 150 features being tracked. We are currently working on a newer version of MuSHR and we recommend using a fair amount of compute power for real-time VI-SLAM (Jetson AGX Xavier or Orin is recommended, but an i7 processor with good amount of RAM and GPU resources would be ideal if your research allows for it). You can always use the MuSHR for collecting sensor data and running the VI-slam on a desktop if you don't need a real-time system.
> 
> That being said, while we haven't tested all of the frameworks listed by you, we have tried using [rtabmap-ros](http://wiki.ros.org/rtabmap_ros#Visual_and_Lidar_Odometry) package for Visual odometry (which internally implements something similar to ORB-SLAM except that the feature detector is interchangeable with FAST/BRIEF /SIFT/SURF/ORB and so on). We tried VINS-Fusion (RGB-D I think) but I believe it had some issues with calibration/initialization which have been faced by [others](https://github.com/HKUST-Aerial-Robotics/VINS-Fusion/issues?q=realsense) as well. Here's a [basic launch file](https://gist.github.com/naughtyStark/9345351fc163c0507adc8cd8e85829ed) that uses rtabmap's rgb-d VO and here is the corresponding [realsense launch file](https://gist.github.com/naughtyStark/de431b70f4017fc8dfd8e6c993554112). I believe rtabmap ros also allows running other state of the art visual odometry frameworks but we haven't really looked into it. Some useful documentation for that has been provided [here](http://wiki.ros.org/rtabmap_ros/Tutorials/StereoOutdoorMapping).
> 
> Let me know if you have further questions! Sidharth Talia.

Thanks. We haven't been responding but we considered your answer !
We managed to make orbslam 3 work in both realtime and offline, but only with the cameras (had some issues with the jetson).
As it is a school project, we will make a report on what we did (and how), we'll post it here if someone's interested in it.

**Reply by sidtalia** [reply]
Would love to see the report when you're done!

**Reply by SashelI** [reply]
> Would love to see the report when you're done!

Hi, the time we had to work  on the project came to an end, so as promised [here](https://github.com/mgodineau/SlaSHR/blob/master/RAPPORT%20FINAL.pdf) is the final report.
Note that it is in french (as we are french students), and was made for our school, so it may contain some sections you're not interested in.

Regards,
Amélie
